### PR TITLE
Custom resolvers + Nested stack support in the transform

### PR DESCRIPTION
**Pull request by mikeparisstuff**
*Issue #, if available:*
#185
#480
#263 
#561 

*Description of changes:*

This PR is not ready to be merged. 

**Remaining work**

- Add CLI integration tests for the full end-to-end CLI -> Transform -> CloudFormation flow with custom stacks included.
- Update amplify docs

**Changes in PR**

- The `splitStack` function that takes a single Template and transforms it into N many connected nested stacks based on a set of regex rules.
  - https://github.com/mikeparisstuff/amplify-cli/blob/5453c52bc7e1923917f8cc8192a2618fd5ca6ee9/packages/graphql-transformer-core/src/util/splitStack.ts#L40
- The `TemplateFormatter` class that turns the TransformContext into a set of deployment assets.
  - https://github.com/mikeparisstuff/amplify-cli/blob/5453c52bc7e1923917f8cc8192a2618fd5ca6ee9/packages/graphql-transformer-core/src/TransformFormatter.ts#L35
- Added the `ctx.putStackMapping` and `ctx.addToStackMapping` methods on the context.
  - Example usage: https://github.com/mikeparisstuff/amplify-cli/blob/969b80c0387ca4b675266a737e1033b44c1684f7/packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts#L97
- Switched to per-request billing in DynamoDB by default.
- Enabling encryption at rest for DynamoDB tables by default.
- Added amplify utility functions to help the CLI manage/upload/deploy api projects managed by the transform.
- Updates all tests to use the new nested stack deployment routine.
- Updated to cloudform 3.5.0
- Updated the amplify-cli to use the new nested stacks support.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts [hunk 1]
```diff
@@ -93,8 +93,20 @@ export class DynamoDBModelTransformer extends Transformer {
...
+        // into their own stack at the end of the transformation.
+        ctx.putStackMapping(
+            `${def.name.value}ModelStack`,
+            [
+                new RegExp(".*" + def.name.value + "Model", 'i'),
+                new RegExp(".*" + def.name.value + "DataSource", 'i'),
+                new RegExp(".*" + def.name.value + "IAMRole", 'i'),
+                new RegExp("^" + def.name.value + "Table", 'i'),
+                // All resolvers except the search resolver.
+                new RegExp("^[^S].*" + def.name.value + "Resolver", 'i'),
```
Comment (mikeparisstuff) [inline packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts]: This regex will change as it blocks some valid configurations.

**Comment by rajababu3** [comment]
@mikeparisstuff is there any update?

**Comment by mikeparisstuff** [comment]
@rajababu3 Yes, I have this working e2e now in my fork if you are interested and want to try it out and offer feedback. The code is still subject to change and there is still a fair amount of testing, documentation, & review work remaining before this gets merged. I am also researching what we can do in order to automatically migrate existing projects to the new world with nested stacks. I can offer more guidance on how to test the fork if you are interested in doing so. LMK.

**Comment by Spenders** [comment]
@mikeparisstuff thanks for this work.
About the migration, I will have to migrate my existing backend.

What I may need:
A process to select everything that has to migrate:
- Functions
- API
- AppSync
- Auth
- ...

I imagine a 3 steps process:
**Selection**
For each item, I should be able to choose which element available in aws will be "migrate".
API : I need to select which Api Gateway endpoints are or should be, after migration, managed with amplify.

AppSync, today my source of truth is directly the #backend. I'd love to tell to grab schema AND resolvers.
...

**Migration**
That's where your magic happened!

**Deployment**
amplify push will override/overwrite what has to be, reconstruct stacks, ...
In the end I will have locally every items in sync with aws. We are not far from creating a pull to complement the push.

**Comment by Spenders** [comment]
hi @mikeparisstuff 
I'd like a little bit of guidance as you propose in order to use this new nested-stacks cli.

Why ?
I have updated directly my AppSync schema on the console. So my local & backend are not synchronous. I can not anymore use amplify push without breaking things. 
So due to this problem I had to add Analytics and Pushnotification without the cli and I'm not sure that everything is fine (I receive push but Eligible endpoints counter in the Segments of Pinpoint is always at 0, even if the I can see that I have 5 endpoints).

Would love to have some help on this topic. If there is no way to recreate the Analytics & Notification from current cli. Let's create from start a new backend with everything.

**Comment by mikeparisstuff** [comment]
@Spenders @rajababu3 I just created this repo which contains a sample project using the new nested stack. The readme has more details for how to try it out. Interested in feedback so let me know if you try it out.

https://github.com/mikeparisstuff/amplify-cli-nested-api-sample

**Comment by rajababu3** [comment]
@mikeparisstuff Thanks for sharing the details, will test and let you know.

**Comment by afuyo** [comment]
@mikeparisstuff  I have large cloudformation template(#185 ) and would be very interested in trying out the fork.  So far, I run into errors after `npm run setup-dev` from the root of directory of your nested-stack  amplify repo.
lerna ERR! npm run build exited 2 in 'amplify-graphql-types-generator'
/.../
node_modules/@types/yargs/index.d.ts(99,106): error TS2304: Cannot find name 'unknown'.

**Comment by afuyo** [comment]
@mikeparisstuff So I am actually able to  deploy a schema that would be  beyond the limit. Good stuff:) Thanks a lot.

**Comment by doom777** [comment]
So DynamoDB capacity will now always be On Demand, or will there be a switch to enable either one? Like an interactive question, or maybe  a --on-demand/--provisioned switch?

**Comment by hakanson** [comment]
Will this also support adding encryption at rest for elasticsearch (via @searchable directive) and adding tags to resources (we have an enterprise tagging standard)?  Should I open that as another issue?

**Comment by kawallis** [comment]
> @mikeparisstuff I have large cloudformation template(#185 ) and would be very interested in trying out the fork. So far, I run into errors after `npm run setup-dev` from the root of directory of your nested-stack amplify repo.
> lerna ERR! npm run build exited 2 in 'amplify-graphql-types-generator'
> /.../
> node_modules/@types/yargs/index.d.ts(99,106): error TS2304: Cannot find name 'unknown'.

Im seeing same error

**Comment by kaustavghosh06** [comment]
Hey guys, you would need to do a `git checkout nested-stack` and then run `npm run setup-dev` for it to work.

**Comment by kawallis** [comment]
Hey that definitely help installing it but im getting this error when trying to run amplify add api
Error: Yaml is not yet supported. Please convert the CloudFormation stack CustomResources.yml to json

**Comment by MatthieuLab** [comment]
Hi @mikeparisstuff 

I have made some tests with your nested stacks branch. And thanks for this work.
I'd like to update you on a bug during the add/creation of my graphql schema push.

I have an error during the creation in the cloud.
Here are the logs from the cli: 

CREATE_FAILED      TenantModelStack                                              AWS::CloudFormation::Stack Mon Jan 21 2019 14:59:48 GMT+0100 (GMT+01:00) Template format error: Unresolved resource dependencies [CenterTenantTable] in the Resources block of the template
CREATE_FAILED      ProductModelStack                                             AWS::CloudFormation::Stack Mon Jan 21 2019 14:59:48 GMT+0100 (GMT+01:00) Template format error: Unresolved resource dependencies [CategoryProductTable] in the Resources block of the template


What are CenterTenantTable and CategoryProductTable ? They are table allowing me to create many-to-many relationship. Here is an extract of my schema: 

`type Center @model {
	id: ID!
	...
	tenants: [CenterTenant] @connection(name: "CenterTenant")
}

type CenterTenant @model {
	id: ID!
	center: Center! @connection(name: "CenterTenant")
	tenant: Tenant! @connection(name: "TenantCenter")
}

type Tenant @model {
	id: ID!
        ...
	centers: [CenterTenant] @connection(name: "TenantCenter")
}`

I have the same kind of code organisation with Category & Product for CategoryProductTable

Did you manage to use many to many relationship (with the recommended intermediate table) with your nested-stacks branch?
Thanks

**Comment by amlcodes** [comment]
How to merge this pull request into my cloned repo?

**Comment by doom777** [comment]
> How to merge this pull request into my cloned repo?

add original as a remote, and merge it into your working branch

**Comment by MatthieuLab** [comment]
Hi guyz,

Just so you know, we are working with this version since one week, updating frequently the cli, and it works like a charm! It solved all my issues with big schema.

2 problems remain, I hope it will help: 

1. I have an authorization Table on DynamoDB, I have to make connections on other tables. When I have more than 5 connections (so 5 tables with "authorizations: [Authorization] @connection" , the cli tell me that I have to many GSI.

Or we can have up to 20 GSI to my understanding.
_Secondary Indexes Per Table
You can define a maximum of 5 local secondary indexes.
There is an initial limit of **20 global secondary indexes** per table. _

I've just created a new GSI from the DynamoDB on my Authorization table and it's OK.

2. More a suggestion than a problem.
I have decided to use and deploy with CloudFormation a stack that can handle image processing. It involve API Gateway, Lambda, CloudFront and S3. Very useful!
[https://aws.amazon.com/fr/answers/web-applications/serverless-image-handler/](https://aws.amazon.com/fr/answers/web-applications/serverless-image-handler/)

I'd love to reference in the aws-export.json this endpoint, event it did not have been created through the amplify-cli.

My suggestion -> Having the ability to select an existing endpoint with the amplify api add. This will allow us to regroup in one file all of our endpoints (in aws_cloud_logic_custom)

Thanks @mikeparisstuff  and all of you for this great work.

**Comment by sakhmedbayev** [comment]
Testing the branch. I am getting `Cannot read property 'AuthCognitoUserPoolId' of undefined` error on `amplify push`

my version is `0.2.2-multienv.0`

**Comment by kaustavghosh06** [comment]
@sakhmedbayev We fixed the issue yesterday. Did you get a chance to pull the code and test it again?

**Comment by sakhmedbayev** [comment]
@kaustavghosh06 thanks for asking. I have been trying to use the "nested-stacks" branch on my ongoing project for several days, but faced many issues and had to go back to the master branch. With the "nested-stacks" branch, I managed to deploy only "fresh" and "simple" schemas. Whenever I tried to add custom stacks "amplify push" failed. :-( I guess, I will wait for the official release. I am very looking forward to it.

**Comment by kawallis** [comment]
heyo does this mean that this is available in the main cli now?

**Comment by jernejcic** [comment]
> heyo does this mean that this is available in the main cli now? 😁

I wish! It looks like it was merged into the `multienv` branch. So now have to wait for that one. Really wish there was a roadmap or they used the Milestones feature so we could get a rough idea on timing... maybe even help move things along.

**Comment by gabriel-wilkes** [comment]
@kawallis @jernejcic Actually this appears to be in the main cli now. They bumped the version to 1.0 and current version is 1.0.5. I am running this now (confirmed with `amplify -v`) and running `amplify env` brings up the appropriate environment commands confirming this is multienv. Also, if you go to the amplify documentation they removed the beta label from Environments & Teams section.

**Comment by jernejcic** [comment]
@gabriel-wilkes Thanks! I literally just came back here to note and celebrate!  🎊

**Comment by aireater** [comment]
> migrat



>  I am also researching what we can do in order to automatically migrate existing projects to the new world with nested stacks.

@mikeparisstuff any progress on the migration? I've been working around the nested stack issue by updating AppSync and DynamoDB outside of Amplify and I would hate to lose that work.

**Comment by kawallis** [comment]
heyo when setting up multiple environments with this branch i am getting this error 
CustomResourcesjson-VST5L9DCSCS2/fa586a50-2f23-11e9-9327-02f46dd00950 was not successfully created: The following resource(s) failed to create: [GetProductLambdaDataSourceRole]. I was following @mikeparisstuff example docs on creating this custom resource

**Comment by kawallis** [comment]
so i was able to get it to go through by add this too my customResources template
```"RoleName": {
					"Fn::Sub": [
						"CreateCustomScanLambdaDataSourceRole-${env}",
						{
							"env": {
								"Ref": "env"
							}
						}
					]
				},```

**Comment by cocacrave** [comment]
Is doc available anywhere?

**Pull request merged**